cachelib/allocator/CacheAllocator.h

```diff
@@ -2115,6 +2099,125 @@ class CacheAllocator : public CacheBase {
     return evictions;
   }
 
+  // exposed for the background evictor to iterate through the memory and evict
+  // in batch. This should improve insertion path for tiered memory config
+  size_t traverseAndEvictItemsUsingDsa(unsigned int tid,
+                                       unsigned int pid,
+                                       unsigned int cid,
+                                       size_t batch) {
+    auto& mmContainer = getMMContainer(tid, pid, cid);
+    size_t evictions = 0;
+    size_t evictionCandidates = 0;
+    std::vector<Item*> candidates;
+    candidates.reserve(batch);
+
+    size_t tries = 0;
+    mmContainer.withEvictionIterator(
+        [&tries, &candidates, &batch, &mmContainer, this](auto&& itr) {
+          while (candidates.size() < batch &&
+                 (config_.maxEvictionPromotionHotness == 0 ||
+                  tries < config_.maxEvictionPromotionHotness) &&
+                 itr) {
+            tries++;
+            Item* candidate = itr.get();
+            XDCHECK(candidate);
+
+            if (candidate->isChainedItem()) {
+              throw std::runtime_error("Not supported for chained items");
+            }
+
+            if (candidate->markMoving(true)) {
+              mmContainer.remove(itr);
+              candidates.push_back(candidate);
+            }
+
+            ++itr;
+          }
+        });
+
+    std::vector<WriteHandle> newItemHandles;
+    newItemHandles.reserve(batch);
+    unsigned int validHandleCnt = 0;
+    for (Item* candidate : candidates) {
+      auto evictedToNext = tryEvictToNextMemoryTier(*candidate, true);
+      newItemHandles.emplace_back(std::move(evictedToNext));
+      if (newItemHandles.back()) {
+        validHandleCnt++;
+        continue;
+      }
+      auto token = createPutToken(*candidate);
+      auto ret = candidate->markExclusiveWhenMoving();
+      XDCHECK(ret);
+      unlinkItemExclusive(*candidate);
+      // wake up any readers that wait for the move to complete
+      // it's safe to do now, as we have the item marked exclusive and
+      // no other reader can be added to the waiters list
+      wakeUpWaiters(*candidate, WriteHandle{});
+
+      if (token.isValid() && shouldWriteToNvmCacheExclusive(*candidate)) {
+        nvmCache_->put(*candidate, std::move(token));
+      }
+    }
+
+    /* Set up batch copy of items using DML */
+    auto sequence =
```

Comment:
I don't think we should put DML-related code directly in CacheAllocator.h. This makes it dependent on DML which is not ideal. It would be better to use a callback or some other abstraction for that.

For now the priority should be to get the best performance so you don't need to change it now. But I think it needs to be changed before merging to develop.